hbase-rsgroup/src/test/java/org/apache/hadoop/hbase/rsgroup/TestRSGroupsAdmin2.java

```diff
@@ -459,4 +466,206 @@ public boolean evaluate() throws Exception {
+              serverHasTableRegions = true;
+              break;
+            }
+          }
+          return !serverHasTableRegions && !rsn.getRegionLocation().equals(ss);
+        }
+        return false;
+      }
+    });
+  }
+
+  private <T> Thread recoverRegionStateThread(T owner, Function<T, List<RegionInfo>> getRegions,
+      RegionStateNode rsn, AtomicBoolean changed){
+    return new Thread(() -> {
+      LOG.info("thread1 start running, will recover region state");
+      long current = System.currentTimeMillis();
+      // wait until there is only left the region we changed state and recover its state.
+      // wait time is set according to the number of max retries, all except failed regions will be
+      // moved in one retry, and will sleep 1s until next retry.
+      while (System.currentTimeMillis() - current <= DEFAULT_MAX_RETRY_VALUE * 1000) {
+        List<RegionInfo> regions = getRegions.apply(owner);
+        LOG.debug("server table region size is:{}", regions.size());
+        assert regions.size() >= 1;
+        // when there is exactly one region left, we can determine the move operation encountered
+        // exception caused by the strange region state.
+        if (regions.size() == 1) {
+          assertEquals(regions.get(0).getRegionNameAsString(),
+              rsn.getRegionInfo().getRegionNameAsString());
+          rsn.setState(RegionState.State.OPEN);
+          LOG.info("set region {} state OPEN", rsn.getRegionInfo().getRegionNameAsString());
+          changed.set(true);
+          break;
+        }
+        sleep(5000);
+      }
+    });
+  }
+
+  @Test
+  public void testFailedMoveWhenMoveServer() throws Exception{
+    String groupName = getGroupName(name.getMethodName());
+    rsGroupAdmin.addRSGroup(groupName);
+    final RSGroupInfo newGroup = rsGroupAdmin.getRSGroupInfo(groupName);
+    Pair<ServerName, RegionStateNode> gotPair = createTableAndSetARegionState(newGroup,
+        10);
+    try{
+      rsGroupAdmin.moveServers(Sets.newHashSet(gotPair.getFirst().getAddress()),
+          newGroup.getName());
+      fail("move servers to group should fail");
+    }catch (IOException e){
+      assertTrue(e.getMessage().contains(
+          gotPair.getSecond().getRegionInfo().getRegionNameAsString()));
+    }
+  }
+
+  @Test
+  public void testFailedMoveWhenMoveTable() throws Exception{
+    final RSGroupInfo newGroup = addGroup(getGroupName(name.getMethodName()), 1);
+    Pair<ServerName, RegionStateNode> gotPair = createTableAndSetARegionState(newGroup, 5);
+    try{
+      rsGroupAdmin.moveTables(Sets.newHashSet(tableName), newGroup.getName());
+      fail("move tables to group should fail");
+    }catch (IOException e){
+      assertTrue(e.getMessage().contains(
+          gotPair.getSecond().getRegionInfo().getRegionNameAsString()));
+    }
+  }
+
+  private Pair<ServerName, RegionStateNode> createTableAndSetARegionState(RSGroupInfo rsGroupInfo,
```

Comment:
Small nit: maybe worth rename to something reflecting its behaviour more clearly, such as: _createTableWithRegionSplitting_?

Reply:
Thanks, I'll change the method name.